popde/adaptive_kde.py

```diff
@@ -0,0 +1,178 @@
+    with variable bandwidth per point
+
+    Example:
+    # Example of 3D KDE and plot for verification
+    mean1, sigma1 = 0.0, 1.0
+    mean2, sigma2 = 0.0, 1.0
+    mean3, sigma3 = 0.0, 1.0
+    n_samples = 1000
+    rndgen = np.random.RandomState(seed=1)
+    sample1 = rndgen.normal(mean1, sigma1, size=n_samples)
+    sample2 = rndgen.normal(mean2, sigma2, size=n_samples)
+    sample3 = rndgen.normal(mean3, sigma3, size=n_samples)
+    sample = np.column_stack((sample1, sample2, sample3)) # shape is (n_points, n_features)
+    # Create and fit the adaptive KDE 
+    kde = AdaptiveBwKDE(sample, dim_names=['x', 'y', 'z'], alpha=0.5, input_transf=None)
+    # Generate grid for plotting
+    xgrid = np.linspace(sample1.min(), sample1.max(), 100)
+    ygrid = np.linspace(sample2.min(), sample2.max(), 100)
+    zgrid = np.linspace(sample3.min(), sample3.max(), 100)
+    XX, YY, ZZ = np.meshgrid(xgrid, ygrid, zgrid)
+    eval_pts = np.column_stack((XX.flatten(), YY.flatten(), ZZ.flatten()))
+
+    # Evaluate the KDE at the grid points
+    density_values = kde.evaluate(eval_pts)
+    density_values = density_values.reshape(XX.shape)
+    print(density_values)
+    """
+    def __init__(self, data, backend='KDEpy', bandwidth=1., dim_names=None, alpha=0.5,  input_transf=None, stdize=False, rescale=None):
+        # Additional parameter for the Wang & Wang formula
+        self.alpha = alpha
+        super().__init__(data, input_transf, stdize, rescale,
+                         backend, bandwidth, dim_names) 
+
+        #Adaptive bandwidth: compute pilot kdevals
+        self.pilot_values = self.evaluate(self.kde_data)
+        #Calculate per-point bandwidths as re-assigning self.bandwidth
+        self.calculate_per_point_bandwidths(self.pilot_values)
+
+    def _local_bandwidth_factor(self, kde_values):
+        """
+        Calculate local bandwidth factor using expression in
+        B. Wang and X. Wang, 2007, DOI: 10.1214/154957804100000000.
+
+        Parameters:
+        -----------
+        kde_values : array-like
+            The KDE values at the data point positions.
+
+        Returns:
+        --------
+        loc_bw_factor : array-like
+           local bandwidth factor for adaptive 
+           bandwidth.
+        """
+        #geometric mean of kde values using scipy
+        g = gmean(kde_values)
+        loc_bw_factor = (kde_values / g) ** self.alpha
+        return loc_bw_factor
+
+    def calculate_per_point_bandwidths(self, pilot_values):
+        """
+        Calculate per-point bandwidths according to
+        B. Wang and X. Wang, 2007, DOI: 10.1214/154957804100000000.
+        Parameters:
+        -----------
+        pilot_values : array-like
+            The pilot KDE values at the data point positions.
+
+        Returns:
+        --------
+        per_point_bandwidths : array-like
+        """
+        # Calculate the inverse of local bandwidth using the provided function
```

Comment:
Comment is too long and inaccurate anyway (we're calculating the inverse bandwidth _factor_, not the invers of the bandwidths)
```suggestion
        # Calculate adaptive inverse bandwidth factors
```

Reply:
removed this line